scripts/gcs_upload.py

```diff
@@ -0,0 +1,578 @@
+    storage_client = create_client()
+    bucket = storage_client.get_bucket(bucket_name)
+    tasks = []
+    for f in files:
+        name = PurePath(f).name
+        blob = bucket.get_blob(gcs_path + "/" + name)
+        target_path = cloud_paths_map[name]
+        if blob.name == target_path:
+            continue
+        logger.info(f"renaming blob from {blob.name} to {target_path}")
+        tasks.append(
+            _rename_blob(bucket=bucket, blob=blob, target_path=target_path)
+        )
+    await asyncio.gather(*tasks)
+
+
+def _gsutil_upload_worker(
+    bucket_name,
+    files,
+    gcs_path,
+    cloud_paths_map,
+    worker_id,
+    boto_options=None,
+):
+    """
+    Uploads a list of filepaths to GCS using gsutil
+    Args:
+        bucket_name (str): name of bucket to upload to
+        files (list): list of absolute filepaths to upload to the bucket
+        gcs_path (str): the cloud storage path to upload the files to
+        cloud_paths_map (dict): a dictionary mapping uploaded file names
+                                (not paths) to their target cloud paths
+        worker_id (int): A unique ID for this worker
+        boto_options (list): list of Boto options to pass to gsutil
+    """
+    options_str = ""
+    if boto_options is not None:
+        sep = " -o "
+        options_str = f"{sep}{sep.join(boto_options)}"
+
+    with tempfile.TemporaryDirectory() as tmpdir:
+        logfile = os.path.join(tmpdir, f"gsutil-cp-worker-{worker_id}-log.log")
+        catfile = os.path.join(tmpdir, f"file-subset-{worker_id}.txt")
+        with open(catfile, "w") as f:
+            for file in files:
+                f.write(file + "\n")
+        cmd = f"cat {catfile} | gsutil -m {options_str} cp -L {logfile} -I gs://{bucket_name}/{gcs_path}"
+        ret = subprocess.run(cmd, shell=True)
+
+        failed_uploads = []
+        if ret.returncode != 0:
+            logger.error(
+                f"Error uploading files, gsutil exit code {ret.returncode}"
+            )
+            failed_uploads.extend(_parse_cp_failures(logfile))
+
+    asyncio.run(_rename_blobs(files, bucket_name, gcs_path, cloud_paths_map))
+
+    return failed_uploads
+
+
+def _python_upload_worker(
+    bucket_name, files, gcs_path, root=None, chunk_size=256 * 1024 * 1024
+):
+    """
+    Uploads a list of filepaths to GCS using the google-cloud-storage Python API
+    Args:
+        bucket_name (str): name of bucket
+        files (list): list of absolute paths to upload to bucket
+        gcs_path (str): google cloud storage path to upload files to
+        root (str): directory shared by files to serve as root directory
+        chunk_size (int): set the blob chunk size (bytes).
+                          Increasing this can speed up transfers.
+    """
+    uploader = GCSUploader(bucket_name)
+    return uploader.upload_files(files, gcs_path, root, chunk_size=chunk_size)
+
+
+def _symlink_duplicate_filenames(filepaths, symlink_dir):
```

Comment:
would it be easier just to prefix everything with unique `uuid`? that way you wouldn't have to detect and resolve collisions.

Reply:
(this wouldn't persist once the files get put into their final locations)

Reply:
that would be simpler for sure

Reply:
on second thought, adding a prefix to every file would mean having to rename all of them, even in the cases where it is not necessary. If this is the case, then there is no advantage to `gsutil` in the case of a flat directory. So maybe it is better to remove the `gsutil` code altogether?

Reply:
Yeah, ask @jtyoung84 what he thinks, but at this point my vote is to remove the `gsutil` code to simplify things.

Reply:
`gsutil` code has been removed